### PR TITLE
DAOS-7812 cart: HLC-based SWIM incarnations

### PR DESCRIPTION
**Pull request by liw**
Current engine incarnations increment from 0. When an engine restarts,
its incarnation resets and increments from 0 again, effectively reusing
previous incarnation numbers. Although a recent improvement to swim
robustness (i.e., when a member finds that others consider itself
suspected or dead, it starts a new incarnation) can mitigate this
problem, it still makes reasoning complex. This patch implements
incarnation numbers using HLCs, which are recovered during engine
restarts, so that the same incarnation number will no longer be reused
by different incarnations.

In an engine, the HLC shall generally be read only after the HLC
recovery in server_init is complete, so that any HLC reading is higher
than all previous ones even across engine restarts. To comply, this
patch moves the HLC recovery ahead of dss_module_init_all (for
vos_mod_init) and dss_srv_init (for crt_swim_init).

Incarnations can help daos order a SWIM DEAD event and a JoinReq, which
is essentially a SWIM ALIVE event. This patch exposes incarnations to
daos via a new crt_self_incarnation_get method and a new incarnation
parameter for crt_event_cb.

Signed-off-by: Li Wei <wei.g.li@intel.com>

### CODE REVIEW

**Comment by liw** [comment]
@dmiter, initial thoughts?

File: src/include/cart/swim.h [hunk 1]
```diff
@@ -128,6 +128,16 @@ struct swim_ops {
 	 */
 	int (*set_member_state)(struct swim_context *ctx,
 				swim_id_t id, struct swim_member_state *state);
+
+	/**
+	 * Increase the incarnation number of a given member.
```
Comment (dmiter) [inline src/include/cart/swim.h:133]: I think "Increase" is not appropriate word for this description. Maybe "Set new incarnation number of a given member." is better?

**Review by dmiter — APPROVED** [review]
This is an interesting idea. I think it should work.

File: src/include/cart/swim.h (hunk 1 above)
Comment (liw) [reply src/include/cart/swim.h:133]: Sure, will change to that.

**Review by liw — COMMENTED** [review]
Thanks. I'll continue with this approach then. Will need to make sure the new crt_hlc_get calls happen _after_ the HLC recovery in server_init (src/engine/init.c). And might also need a way for engine to get the initial self incarnation and report to the MS, who currently has trouble ordering Joins and exclusions.

**Comment by dmiter** [comment]
I'm only worry about initial incarnation numbers which will be different on each node. How they will sync? Probably this will be the issue.

**Comment by liw** [comment]
> I'm only worry about initial incarnation numbers which will be different on each node. How they will sync? Probably this will be the issue.

Suppose we have 3 engines. Initially, they look like this:

    rank 0:
      0 at incarnation 1234
      1 at incarnation 0
      2 at incarnation 0

    rank 1:
      0 at incarntion 0
      1 at incarnation 1230
      2 at incarnation 0

    rank 2:
      0 at incarnation 0
      1 at incarnation 0
      2 at incarnation 1235

Then, if rank 0 recevies "rank 2 ALIVE incarnation 1235", we'll have:

    rank 0:
      0 at incarnation 1234
      1 at incarnation 0
      2 at incarnation 1235

    rank 1:
      0 at incarntion 0
      1 at incarnation 1230
      2 at incarnation 0

    rank 2:
      0 at incarnation 0
      1 at incarnation 0
      2 at incarnation 1235

So rank 0's knowledge about rank 2 has been synchronized. Correct?

**Comment by dmiter** [comment]
Ahh, yes, you are right. I'm not realized you set incarnation number only for self initially.

**Comment by mjmac** [comment]
FYI, I am testing this in #6266 and it seems to be the missing piece I've been looking for. The test I added is intended to verify that the replicated MS behaves as expected in the face of various failure/recovery scenarios. Without your PR as a base, the tests fail intermittently due to stale events. With your PR, I am seeing stale events filtered successfully, e.g.
```
DEBUG 18:49:59.116746 membership.go:499: ignoring rank dead event for previous incarnation of 0 (3edc07e1e600000 < 3edc0d150280000)
```

**Comment by liw** [comment]
Glad to hear that, @mjmac. Thanks.

**Comment by liw** [comment]
osa offline drain: DAOS-8163

**Comment by mjmac** [comment]
@dmiter, @frostedcmos: Just FYI, this PR is now a blocker for 2.0 TB4, as it turns out it is needed for the MS resiliency feature.

**Review by dmiter — APPROVED** [review]
looks good for me.

**Comment by liw** [comment]
Merely merged with the latest master; no changes. The previous build failures were unrelated to this patch.

**Review by frostedcmos — APPROVED** [review]

**Pull request merged**